File: VB/WindowsFormsApplication3/InplaceDataNavigator/RepositoryItemInplaceDataNavigator.vb
Imports Microsoft.VisualBasic
Imports DevExpress.XtraEditors.Registrator
Imports DevExpress.XtraEditors.Repository
Imports DevExpress.XtraEditors.ViewInfo
Imports System
Imports System.Collections.Generic
Imports System.Linq
Imports System.Windows.Forms

Namespace WindowsFormsApplication3
	<System.ComponentModel.DesignerCategory(""), UserRepositoryItem("Register")> _
	Public Class RepositoryItemInplaceDataNavigator
		Inherits RepositoryItem
		Private _editNavigator As MyDataNavigator
		Public Property EditNavigator() As MyDataNavigator
			Get
				Return _editNavigator
			End Get
			Set(ByVal value As MyDataNavigator)
				_editNavigator = value
			End Set
		End Property

		Friend Const EditorName As String = "InplaceDataNavigator"

		Shared Sub New()
			Register()
		End Sub
		Public Sub New()
			AutoHeight = False
			EditNavigator = New MyDataNavigator()
			EditNavigator.Dock = DockStyle.Fill
			EditNavigator.BindingContext = New BindingContext()
		End Sub

		Protected Overrides Sub Finalize()
			EditNavigator.Dispose()
		End Sub

		Public Shared Sub Register()
			EditorRegistrationInfo.Default.Editors.Add(New EditorClassInfo(EditorName, GetType(InplaceDataNavigator), GetType(RepositoryItemInplaceDataNavigator), GetType(BaseEditViewInfo), New InplaceDataNavigatorPainter(), True))
		End Sub

		Public Overrides ReadOnly Property EditorTypeName() As String
			Get
				Return EditorName
			End Get
		End Property

		Public Overrides Sub Assign(ByVal item As RepositoryItem)
			BeginUpdate()
			Try
				MyBase.Assign(item)
				Dim source As RepositoryItemInplaceDataNavigator = TryCast(item, RepositoryItemInplaceDataNavigator)
				If source Is Nothing Then
					Return
				End If
				EditNavigator.DataSource = source.EditNavigator.DataSource
			Finally
				EndUpdate()
			End Try
		End Sub
	End Class
End Namespace

File: CS/WindowsFormsApplication3/InplaceDataNavigator/RepositoryItemInplaceDataNavigator.cs
using DevExpress.XtraEditors.Registrator;
using DevExpress.XtraEditors.Repository;
using DevExpress.XtraEditors.ViewInfo;
using System;
using System.Collections.Generic;
using System.Linq;
using System.Windows.Forms;

namespace WindowsFormsApplication3
{
    [System.ComponentModel.DesignerCategory("")]
    [UserRepositoryItem("Register")]
    public class RepositoryItemInplaceDataNavigator : RepositoryItem
    {
        private MyDataNavigator _editNavigator;
        public MyDataNavigator EditNavigator
        {
            get { return _editNavigator; }
            set { _editNavigator = value; }
        }

        internal const string EditorName = "InplaceDataNavigator";

        static RepositoryItemInplaceDataNavigator() { Register(); }
        public RepositoryItemInplaceDataNavigator()
        {
            AutoHeight = false;            
            EditNavigator = new MyDataNavigator();
            EditNavigator.Dock = DockStyle.Fill;
            EditNavigator.BindingContext = new BindingContext();
        }

        ~RepositoryItemInplaceDataNavigator()
        {
            EditNavigator.Dispose();
        }

        public static void Register()
        {
            EditorRegistrationInfo.Default.Editors.Add(new EditorClassInfo(EditorName, typeof(InplaceDataNavigator),
                typeof(RepositoryItemInplaceDataNavigator), typeof(BaseEditViewInfo),
                new InplaceDataNavigatorPainter(), true));
        }

        public override string EditorTypeName
        {
            get { return EditorName; }
        }

        public override void Assign(RepositoryItem item)
        {
            BeginUpdate();
            try
            {
                base.Assign(item);
                RepositoryItemInplaceDataNavigator source = item as RepositoryItemInplaceDataNavigator;
                if (source == null) return;
                EditNavigator.DataSource = source.EditNavigator.DataSource;
            }
            finally
            {
                EndUpdate();
            }
        }
    }
}
